File: spotledger_hr/api/employee_qr_code.py
import frappe
from frappe import _
from spotledger_hr.utilities.qr_code_generator import get_employee_qr_code, get_employee_qr_code_with_legacy_code


@frappe.whitelist()
def get_employee_qr_code_api(employee_id, use_legacy_code=False):
    """API endpoint to get QR code for an employee.
    
    Args:
        employee_id (str): Employee document name/ID
        use_legacy_code (bool): Whether to use custom_old_code if available
        
    Returns:
        dict: Response with QR code data URL
    """
    try:
        # Validate employee exists
        if not frappe.db.exists("Employee", employee_id):
            frappe.throw(_("Employee {0} not found").format(employee_id))
        
        # Generate QR code
        if use_legacy_code:
            qr_code_data = get_employee_qr_code_with_legacy_code(employee_id)
        else:
            qr_code_data = get_employee_qr_code(employee_id)
        
        return {
            "status": "success",
            "qr_code": qr_code_data,
            "employee_id": employee_id
        }
        
    except Exception as e:
        frappe.logger().error(f"Employee QR Code API error: {str(e)}")
        return {
            "status": "error",
            "message": str(e)
        }


@frappe.whitelist()
def generate_qr_code_for_print(employee_id, use_legacy_code=False):
    """Generate QR code for print format.
    
    Args:
        employee_id (str): Employee document name/ID
        use_legacy_code (bool): Whether to use custom_old_code if available
        
    Returns:
        str: QR code data URL for print format
    """
    try:
        if use_legacy_code:
            return get_employee_qr_code_with_legacy_code(employee_id)
        else:
            return get_employee_qr_code(employee_id)
            
    except Exception as e:
        frappe.logger().error(f"Print QR Code generation error: {str(e)}")
        return None

File: spotledger_hr/public/js/employee_qr_code.js
// Employee QR Code functionality
frappe.ui.form.on('Employee', {
    refresh: function(frm) {
        // Add QR Code generation button
        if (frm.doc.name) {
            frm.add_custom_button(__('Generate QR Code'), function() {
                generate_qr_code(frm);
            }, __('Actions'));
            
            // Auto-generate QR code on form load
            if (!frm.doc.custom_qr_code) {
                generate_qr_code(frm);
            }
        }
    },
    
    custom_old_code: function(frm) {
        // Regenerate QR code when old code changes
        if (frm.doc.name && frm.doc.custom_old_code) {
            generate_qr_code(frm, true);
        }
    }
});

function generate_qr_code(frm, use_legacy_code = false) {
    frappe.call({
        method: 'spotledger_hr.api.employee_qr_code.get_employee_qr_code_api',
        args: {
            employee_id: frm.doc.name,
            use_legacy_code: use_legacy_code
        },
        callback: function(r) {
            if (r.message && r.message.status === 'success') {
                // Update the QR code field with the generated image
                frm.set_value('custom_qr_code', 
                    `<img src="${r.message.qr_code}" alt="QR Code" style="max-width: 200px; height: auto;" />`
                );
                frm.save();
                frappe.show_alert({
                    message: __('QR Code generated successfully'),
                    indicator: 'green'
                });
            } else {
                frappe.msgprint(__('Error generating QR Code: ') + (r.message.message || 'Unknown error'));
            }
        },
        error: function(r) {
            frappe.msgprint(__('Error generating QR Code'));
        }
    });
}


